### PR TITLE
Fixes #8 - Feature/world

### LINKED ISSUE #8: Create a class for world
elements needed 
- boundary
- obstacles
- robot state
- start 
- goal 

functions 
- get state
- update state

### LINKED ISSUE #9: unittest
Evaluate using UnnitTest and give detailed report of the findings

### PR DESCRIPTION
**Pull request by 00arun00**
- updated init to fix imports
- temporary txt map loader (need to change this definitely)
- restructured the project
- partial implementation of world envs
- added a sample txt map file
- added unnitests

### CODE REVIEW

File: .github/workflows/python-package.yml [hunk 1]
```diff
@@ -15,7 +15,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: [3.5, 3.6, 3.7, 3.8]
+        python-version: [3.6, 3.7, 3.8]
```
Comment (00arun00) [inline .github/workflows/python-package.yml]: had to remove it since dataclasses are not supported <3.6

File: pybotic/__init__.py [hunk 2]
```diff
@@ -1 +1,2 @@
-from pybotic import world
+from . import worlds
+from . import utils
```
Comment (00arun00) [inline pybotic/__init__.py:2]: fix importing

File: pybotic/utils/__init__.py [hunk 3]
```diff
@@ -0,0 +1 @@
+from . import world_utils
```
Comment (00arun00) [inline pybotic/utils/__init__.py:1]: fix importing

File: pybotic/utils/world_utils.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
+import numpy as np
+import os.path
+
+
+def load_3d_world_map(file_name):
```
Comment (00arun00) [inline pybotic/utils/world_utils.py:5]: just temp we will re-write this

File: pybotic/worlds.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        if np.shape(self._goal) not in {(3, 1), (3,)}:
+            raise ValueError(f'wrong shape {np.shape(self._goal)}')
+
+        if np.shape(self._boundary) not in {(6, 1), (6,)}:
+            raise ValueError(f"wrong shape {np.shape(self._boundary)}")
+
+        for obs_idx in self._obstacles:
+            shape_ = np.shape(self._obstacles[obs_idx])
+            if shape_ not in {(6, 1), (6,)}:
+                raise ValueError(f"wrong shape {shape_}")
```
Comment (00arun00) [inline pybotic/worlds.py]: validated the inputs send to the constructor, __post_init__() runs automatically after auto-generated __init__()

File: pybotic/worlds.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+            raise TypeError(err_string)
+
+        self._robot_pose = robot_action
+
+    @abstractmethod
+    def render(self):
+        """
+            renders the envirnment
+        """
+        # can't have pass here for coverage reasons
```
Comment (00arun00) [inline pybotic/worlds.py:90]: can't add pass statement here since it will cause problems with coverage so removing it

File: pybotic/worlds.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+            Args:
+                f_name: (str) path to file
+
+            returns:
+
+        """
+        start, goal, obstacles, boundary = load_3d_world_map(f_name)
+        if start is None:
+            start = np.zeros((3, 1))
+        return cls(boundary, obstacles, start, goal)
```
Comment (00arun00) [inline pybotic/worlds.py]: currently will error out with the current load_3d_world_map() will be fine after that is fixed

File: setup.py [hunk 8]
```diff
@@ -14,7 +14,6 @@
         'License :: OSI Approved :: MIT License',
         'Framework :: Robot Framework',
         'Programming Language :: Python :: 3',
-        'Programming Language :: Python :: 3.5',
```
Comment (00arun00) [inline setup.py:17]: remove support

File: tests/sample_world.txt [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+block  0.0  2.0  4.5 10.0  2.5  6.0 120 120 120
+
+block  0.0  2.0  1.5  3.0  2.5  4.5 120 120 120
+block  7.0  2.0  1.5 10.0  2.5  4.5 120 120 120
+
+block  3.0  0.0  2.4  7.0  0.5  4.5 120 120 120
+
+block 0.0 15.0  0.0 10.0 20.0  1.0 120 120 120
+block 0.0 15.0  1.0 10.0 16.0  3.5 120 120 120
+block 0.0 18.0  4.5 10.0 19.0  6.0 120 120 120
```
Comment (00arun00) [inline tests/sample_world.txt:17]: sample not frozen can be used for rewriting load_3d_map() 

- probably need to add start and goal to it

File: tests/test_Continous3D_Static.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
+from pybotic.worlds import Continous3D_Static
+import unittest
+import numpy as np
+
+
+class TestContinous3DStatic(unittest.TestCase):
```
Comment (00arun00) [inline tests/test_Continous3D_Static.py:8]: testing continous3dstatic 
Missing tests for :
- create_from_txt_file()

File: pybotic/worlds.py (hunk 7 above)
Comment (00arun00) [inline pybotic/worlds.py]: tests are missing currently

File: pybotic/worlds.py (hunk 5 above)
Comment (gunjanpatil) [reply pybotic/worlds.py]: add automated conversion to suggested shapes and types

File: pybotic/worlds.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+@dataclass
+class World(ABC):
+    """
+        This is an abstract class representing world
+        Never to be directly used in any applications
+    """
+    _boundary: np.ndarray = field(default=None)
+    _obstacles: Dict[str, np.ndarray] = field(default=None)
+    _start: np.ndarray = field(default=None)
+    _goal: np.ndarray = field(default=None)
```
Comment (gunjanpatil) [inline pybotic/worlds.py]: Change type from np.ndarray to node based structure

File: pybotic/worlds.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+            Returns:
+                A dictionary containg.__annotations__
+                boundary: (numpy.ndarray) boundary of the world
+
+                start: (numpy.ndarray) start location
+                goal: (numpy.ndarray) goal location
+                robot_location: (numpy.ndarray) location of the
+                    robot
+                Obstacle_state: (numpy.ndarray) state/location of obstacles
```
Comment (gunjanpatil) [inline pybotic/worlds.py]: Fix the order of keys returned in the dict

File: pybotic/worlds.py [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+        """
+        if np.shape(robot_action) not in {(3, 1), (3,)}:
+            raise ValueError('wrong shape')
+
+        if not isinstance(robot_action, type(self._robot_pose)):
+            err_string = (f"expected {type(self._robot_pose)}"
+                          f" got {type(robot_action)}")
+            raise TypeError(err_string)
+
+        self._robot_pose = robot_action
```
Comment (gunjanpatil) [inline pybotic/worlds.py]: add automated conversion to suggested shapes and types

File: pybotic/worlds.py [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+
+    def __call__(self):
+        return self.get_state()
+
+    def update_state(self, robot_action):
+        """
+            currently just pass through
+
+            Args:
+                robot_action: (iterable 3d) robot pose
```
Comment (gunjanpatil) [inline pybotic/worlds.py]: fix args type

File: pybotic/worlds.py (hunk 6 above)
Comment (00arun00) [reply pybotic/worlds.py:90]: https://www.codependentcodr.com/code-quality-challenge-1-test-coverage.html

File: pybotic/worlds.py [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+            yield name[1:], self.__dict__[name]
+
+
+@dataclass
+class Continous3D_Static(World):
+    """
+        Implementation of a continous 3D world
+    """
+
+    @classmethod
```
Comment (gunjanpatil) [inline pybotic/worlds.py]: fix inheritance

File: tests/test_Continous3D_Static.py [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+        self.boundary = np.array([1, 2, 3, 4, 5, 6])
+        self.obstacles = {'1': np.array([1]*6)}
+        self.start = np.zeros((3, 1))
+        self.goal = np.ones((3, 1))
+        self.c = Continous3D_Static(self.boundary, self.obstacles,
+                                    self.start, self.goal)
+
+    def test_valid(self):
+        """
+            this is will check all the valid cases
```
Comment (gunjanpatil) [inline tests/test_Continous3D_Static.py]: fix grammar

File: tests/test_Continous3D_Static.py [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+
+class TestContinous3DStatic(unittest.TestCase):
+    """
+        Tester for Continous3D_Static
+        test covered:
+            - valid
+            - file not found
+    """
+
+    def setUp(self):
```
Comment (gunjanpatil) [inline tests/test_Continous3D_Static.py:16]: add docstring

File: tests/test_Continous3D_Static.py [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+        self.boundary = np.array([1, 2, 3, 4, 5, 6])
+        self.obstacles = {'1': np.array([1]*6)}
+        self.start = np.zeros((3, 1))
+        self.goal = np.ones((3, 1))
+        self.c = Continous3D_Static(self.boundary, self.obstacles,
+                                    self.start, self.goal)
```
Comment (gunjanpatil) [inline tests/test_Continous3D_Static.py]: Test constructor

File: tests/test_Continous3D_Static.py [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+
+    def test_valid(self):
+        """
+            this is will check all the valid cases
+            - valid construction
+            - get state
+
+        """
+        # valid construction
+        self.assertIsInstance(self.c, Continous3D_Static)
```
Comment (gunjanpatil) [inline tests/test_Continous3D_Static.py]: test constructor

File: tests/test_Continous3D_Static.py [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+    def rec_check_dict(self, a, b):
+        # check keys
+        assert a.keys() == b.keys()
+
+        # ensure that values are same
+        for key in a.keys():
+            if isinstance(a[key], dict):
+                self.rec_check_dict(a[key], b[key])
+                continue
+            assert all(np.equal(a[key], b[key]))
```
Comment (gunjanpatil) [inline tests/test_Continous3D_Static.py]: move to test utils

File: tests/test_Continous3D_Static.py [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+                        'start': self.start,
+                        'goal': self.goal,
+                        'robot_pose': self.start}
+
+        self.rec_check_dict(valid_output, self.c())
+
+    def test_empty(self):
+        with self.assertRaises(ValueError):
+            Continous3D_Static()
+
```
Comment (gunjanpatil) [inline tests/test_Continous3D_Static.py]: comments

File: tests/test_Continous3D_Static.py (hunk 20 above)
Comment (gunjanpatil) [reply tests/test_Continous3D_Static.py]: use unit test aserts

**Review by gunjanpatil — CHANGES_REQUESTED** [review]

File: pybotic/worlds.py (hunk 5 above)
Comment (gunjanpatil) [reply pybotic/worlds.py]: Allow empty obstacles

File: tests/test_Continous3D_Static.py (hunk 20 above)
Comment (00arun00) [reply tests/test_Continous3D_Static.py]: removed since no longer needed

File: tests/test_Continous3D_Static.py (hunk 19 above)
Comment (00arun00) [reply tests/test_Continous3D_Static.py]: done

File: tests/test_Continous3D_Static.py (hunk 18 above)
Comment (00arun00) [reply tests/test_Continous3D_Static.py]: done

File: tests/test_Continous3D_Static.py (hunk 17 above)
Comment (00arun00) [reply tests/test_Continous3D_Static.py:16]: done

File: pybotic/worlds.py [hunk 22]
```diff
@@ -0,0 +1,154 @@
...
+from pybotic.geometry import Point3D, Cuboid
+
+
+@dataclass
+class World(ABC):
+    """Abstaract World class
+
+    This is an abstract class representing world
+    Never to be directly used in any applications
+    """
```
Comment (gunjanpatil) [inline pybotic/worlds.py]: add comments for args

File: pybotic/worlds.py [hunk 23]
```diff
@@ -0,0 +1,154 @@
...
+from dataclasses import dataclass, field
+import numpy as np
+from typing import Dict
+from typeguard import typechecked, check_type
+
+from pybotic.utils.world_utils import load_3d_world_map
+from pybotic.geometry import Point3D, Cuboid
+
+
+@dataclass
```
Comment (gunjanpatil) [inline pybotic/worlds.py:11]: make class frozen

File: pybotic/worlds.py [hunk 24]
```diff
@@ -0,0 +1,154 @@
...
+
+        Args:
+            f_name (str): path to file
+
+        Returns:
+            object (Continous3D_Static): object of class
+        """
+        boundary, obstacles, start, goal = load_3d_world_map(f_name)
+        if start is None:
+            start = Point3D.create_from_iter(np.zeros((3, 1)))
```
Comment (gunjanpatil) [inline pybotic/worlds.py]: no need

**Review by gunjanpatil — CHANGES_REQUESTED** [review]

**Review by gunjanpatil — APPROVED** [review]

**Pull request merged**